ddl/partition.go

```diff
@@ -280,3 +281,97 @@ func getPartitionIDs(table *model.TableInfo) []int64 {
 	}
 	return partitionIDs
 }
+
+// checkRangePartitioningKeysConstraints checks that the range partitioning key is included in the table constraint.
+func checkRangePartitioningKeysConstraints(ctx sessionctx.Context, expr ast.ExprNode, tblInfo *model.TableInfo, constraints []*ast.Constraint) error {
+	// Returns directly if there is no constraint in the partition table.
+	if len(constraints) == 0 {
+		return nil
+	}
+
+	// Save the column names in the table constraint in slice.
+	uniKeys, multipleKeys, priKeys := saveConstraintsColumnNames(tblInfo, constraints)
+
+	// Parse partitioning key, save the column names in the partitioning key to slice.
+	buf := new(bytes.Buffer)
+	expr.Format(buf)
+	var partkeys []string
+	e, err := expression.ParseSimpleExpr(ctx, buf.String(), tblInfo)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	cols := expression.ExtractColumns(e)
+	for _, col := range cols {
+		partkeys = append(partkeys, col.ColName.L)
+	}
+
+	switch expr.(type) {
+	case *ast.ColumnNameExpr, *ast.FuncCallExpr:
+		// Only one column name expression.
```

Comment:
This make me confused.

```
partition by range (year(col1) + year(col2))
```

Is it `*ast.FuncCallExpr` and is there "only one column name expression" ?